src/cli.rs

```diff
@@ -0,0 +1,764 @@
+use std::path::{Path, PathBuf};
+use std::time::Duration;
+
+#[cfg(feature = "completions")]
+use anyhow::anyhow;
+use clap::{
+    builder::RangedU64ValueParser, value_parser, AppSettings, Arg, ArgAction, ArgGroup, ArgMatches,
+    Command, ErrorKind, Parser, ValueEnum,
+};
+#[cfg(feature = "completions")]
+use clap_complete::Shell;
+use normpath::PathExt;
+
+use crate::error::print_error;
+use crate::exec::CommandSet;
+use crate::filesystem;
+#[cfg(unix)]
+use crate::filter::OwnerFilter;
+use crate::filter::SizeFilter;
+
+// Type for options that don't have any values, but are used to negate
+// earlier options
+struct Negations;
+
+impl clap::FromArgMatches for Negations {
+    fn from_arg_matches(_: &ArgMatches) -> clap::Result<Self> {
+        Ok(Negations)
+    }
+
+    fn update_from_arg_matches(&mut self, _: &ArgMatches) -> clap::Result<()> {
+        Ok(())
+    }
+}
+
+impl clap::Args for Negations {
+    fn augment_args(cmd: Command<'_>) -> Command<'_> {
+        Self::augment_args_for_update(cmd)
+    }
+
+    fn augment_args_for_update(cmd: Command<'_>) -> Command<'_> {
+        cmd.arg(
+            Arg::new("no-hidden")
+                .long("no-hidden")
+                .overrides_with("hidden")
+                .hide(true)
+                .long_help("Overrides --hidden."),
+        )
+        .arg(
+            Arg::new("ignore")
+                .long("ignore")
+                .overrides_with("no-ignore")
+                .hide(true)
+                .long_help("Overrides --no-ignore."),
+        )
+        .arg(
+            Arg::new("ignore-vcs")
+                .long("ignore-vcs")
+                .overrides_with("no-ignore-vcs")
+                .hide(true)
+                .long_help("Overrides --no-ignore-vcs."),
+        )
+        .arg(
+            Arg::new("relative-path")
+                .long("relative-path")
+                .overrides_with("absolute-path")
+                .hide(true)
+                .long_help("Overrides --absolute-path."),
+        )
+        .arg(
+            Arg::new("no-follow")
+                .long("no-follow")
+                .overrides_with("follow")
+                .hide(true)
+                .long_help("Overrides --follow."),
+        )
+    }
+}
```

Comment:
Its not clear to me why this isn't using the derive API?  Just to avoid having the fields?

Reply:
> Just to avoid having the fields?

I think that was probably my motivation. These options are all just to negate other options, so having fields for them wasn't terribly useful. But maybe having a few unused boolean fields wouldn't be too bad.

Reply:
I wonder if it'd be worth supporting `()` type for cases like this in clap derive

Reply:
For now, I think I prefer this solution over the version with duplicated booleans in the struct.